KAFKA-10547; add topicId in MetadataResp

dengziming opened the pull request:
*More detailed description of your change*

1. Bump the version of MetadataReq and MetadataResp, add topicId in MetadataResp
2. Alter describeTopic in AdminClientTopicService and ZookeeperTopicService
3. TopicMetadata is cached in MetadataCache, so we need to add topicId to MetadataCache
4. MetadataCache is updated by UpdateMetadataRequest, bump the version of UpdateMetadataReq and UpdateMetadataResp, add topicId in UpdateMetadataReq.

*Summary of testing strategy (including rationale)*

Tested locally, here is some result:

New server + new Client : 

kafka-topics.sh --describe --zookeeper localhost:2181 --topic old-version-topic 
Topic: old-version-topic	TopicId: wRPl6VAlQeyE77bDxEESzg	PartitionCount: 2	ReplicationFactor: 1	Configs: 
	Topic: old-version-topic	Partition: 0	Leader: 0	Replicas: 0	Isr: 0
	Topic: old-version-topic	Partition: 1	Leader: 0	Replicas: 0	Isr: 0

kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic old-version-topic 
Topic: old-version-topic	TopicId: wRPl6VAlQeyE77bDxEESzg	PartitionCount: 2	ReplicationFactor: 1	Configs: segment.bytes=1073741824
	Topic: old-version-topic	Partition: 0	Leader: 0	Replicas: 0	Isr: 0
	Topic: old-version-topic	Partition: 1	Leader: 0	Replicas: 0	Isr: 0

Old Server + new Client

kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic old-version-topic 
Topic: old-version-topic	PartitionCount: 2	ReplicationFactor: 1	Configs: segment.bytes=1073741824
	Topic: old-version-topic	Partition: 0	Leader: 0	Replicas: 0	Isr: 0
	Topic: old-version-topic	Partition: 1	Leader: 0	Replicas: 0	Isr: 0

New server + old client

kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic old-version-topic 
Topic: old-version-topic	PartitionCount: 2	ReplicationFactor: 1	Configs: segment.bytes=1073741824
	Topic: old-version-topic	Partition: 0	Leader: 0	Replicas: 0	Isr: 0
	Topic: old-version-topic	Partition: 1	Leader: 0	Replicas: 0	Isr: 0


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dengziming commented:
@rajinisivaram @jolshan Hi, PTAL.

jolshan commented:
Hi @dengziming! Thanks for the PR! I was hoping to add LeaderAndIsrRequests before UpdateMetadata/Metadata, as ordered in the [JIRA ticket.](https://issues.apache.org/jira/browse/KAFKA-10545) There are just a few features for persisting the topic IDs I wanted to include. I'm thinking we could review this PR and my PR: https://github.com/apache/kafka/pull/9626 at the same and merge mine first and yours immediately after.

jolshan commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -327,6 +350,18 @@ public Node controller() {
+        return topicIds.getOrDefault(topic, Uuid.ZERO_UUID);
+    }
+
+    public String getTopicName(Uuid topiId) {
```
nit: `Uuid topicId`

jolshan replied:
Also, for getTopicId, we do getOrDefault. Is there a reason we wouldn't do that here?

dengziming replied:
Here, I try to use getOrDefault but don't know what the default value is(it seems that "" is not a good default topicName). so just return null and let the caller decide, suggestions are welcomed if you have good ideas.

jolshan replied:
I see...maybe I missed it, but I did not see the method used. What use case would you expect for topic names in Cluster?

dengziming replied:
This is not used so far but will be used when we get TopicMetadata by topicId, so just add it in advance.

jolshan commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -113,6 +114,7 @@ object TopicCommand extends Logging {
     def printDescription(): Unit = {
       val configsAsString = config.entries.asScala.filter(!_.isDefault).map { ce => s"${ce.name}=${ce.value}" }.mkString(",")
       print(s"Topic: $topic")
+      print(s"\tTopicId: ${if(topicId == Uuid.ZERO_UUID) "" else topicId}")
```
Do we want to print an empty string for TopicId? Would it make sense to include the Zero Uuid or not print a topic ID label at all?

jolshan commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -424,6 +426,13 @@ case object KAFKA_2_7_IV2 extends DefaultApiVersion {
   val id: Int = 30
 }
 
+case object KAFKA_2_7_IV3 extends DefaultApiVersion {
```
nit: this should be KAFKA_2_8_IV0, or KAFKA_2_8_IV1 if https://github.com/apache/kafka/pull/9601 merges

jolshan commented:
@dengziming Looks pretty good so far to me! I think it would be useful to write a few unit/integration tests to ensure the metadata snapshot behavior and describe topics work as expected.

jolshan commented on clients/src/main/resources/common/message/MetadataRequest.json:
```diff
@@ -31,9 +31,11 @@
       "about": "The topics to fetch metadata for.", "fields": [
-      { "name": "Name", "type": "string", "versions": "0+", "entityType": "topicName",
+      { "name": "TopicId", "type": "uuid", "versions": "10+", "ignorable": true, "about": "The topic id." },
+      { "name": "Name", "type": "string", "versions": "0+", "entityType": "topicName", "ignorable": true,
```
As discussed in the mailing thread and the KIP writeup, this should be "nullable" rather than ignorable

dengziming replied:
I tried to use nullable but got an Exceptioin with `Unrecognized field "nullable"` message.

dengziming commented:
> @dengziming Looks pretty good so far to me! I think it would be useful to write a few unit/integration tests to ensure the metadata snapshot behavior and describe topics work as expected.

I added/altered some unit/integration tests in `MetadataRequestTest` and `ControllerChannelManagerTest`, take a look.

Commit:
KAFKA-10547; alter some unit tests

KAFKA-10547; alter some integration tests

KAFKA-10547; optimeze imports

resolve comments

rajinisivaram commented on core/src/main/scala/kafka/log4j.properties:
```diff
@@ -0,0 +1,91 @@
+
+# Unspecified loggers and loggers with additivity=true output to server.log and stdout
+# Note that INFO only applies to unspecified loggers, the log level of the child logger is used otherwise
+log4j.rootLogger=INFO, stdout, kafkaAppender
```
Was this file checked in by mistake? Can we remove?

rajinisivaram commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -372,6 +380,8 @@ class MetadataCache(brokerId: Int) extends Logging {
 
   case class MetadataSnapshot(partitionStates: mutable.AnyRefMap[String, mutable.LongMap[UpdateMetadataPartitionState]],
+                              topicIds: Map[String, Uuid],
+                              topicNames: Map[Uuid, String],
```
It seems unnecessary to include both topicIds and topicNames in the constructor of this case class. We can add `val topicNames` inside this class that creates the second map from the first.

rajinisivaram replied:
Also same question as for Cluster: Do we have a use case that requires both maps to be stored for fast lookups? Or would it be sufficient to store it one-way and look up the single map for the other way?

rajinisivaram commented on core/src/test/scala/unit/kafka/controller/ControllerChannelManagerTest.scala:
```diff
@@ -873,6 +883,11 @@ class ControllerChannelManagerTest {
     }
+
+    context.allTopics ++= topics
+    for (topic <- topics if topicIds.contains(topic)) {
```
Couldn't we just do this instead of the for loop:
```
 topicIds.foreach { case (name, id) => context.addTopicId(name, id) }
```

dengziming replied:
done!

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -46,6 +46,8 @@
     private final Map<Integer, Node> nodesById;
     private final ClusterResource clusterResource;
+    private final Map<String, Uuid> topicIds;
+    private final Map<Uuid, String> topicNames;
```
Do we have a use case that requires both maps to be stored in Cluster for fast lookups? Or would it be sufficient to store it one-way and look up the single map for the other way?

dengziming replied:
Thank you for your suggestions, the `topicNames` is not used so I just removed it. Maybe we will add them if it's necessary in a future use case.

rajinisivaram commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -314,9 +315,16 @@ class MetadataCache(brokerId: Int) extends Logging {
+        .filter(_._2 != Uuid.ZERO_UUID).toMap
+      val topicIds = mutable.Map.empty[String, Uuid]
+      topicIds.addAll(metadataSnapshot.topicIds)
+      topicIds.addAll(newTopicIds)
```
We seem to be adding to the topicIds/topicNames map, but never removing anything.

dengziming replied:
Hi, @rajinisivaram ,Thank you for your reply. I find that when KafkaController deletes a topic it will send `updateMetadataRequest` with all the partitions in it, and set `state.leader=-2` indicating the partition is deleted. And the `MetadataCache` will not remove the topic from cache but just keep the `UpdateMetadataPartitionState` of the deleted topics, so I also keep the topicId when deleted.

here I tried to delete a topic and then describe it.

➜  ~ kafka-topics.sh --delete --zookeeper localhost:2181 --topic old-version-topic
Topic old-version-topic is marked for deletion.
Note: This will have no impact if delete.topic.enable is not set to true.

➜  ~ kafka-topics.sh --describe --zookeeper localhost:2181 --topic old-version-topic
Topic: old-version-topic TopicId: wRPl6VAlQeyE77bDxEESzg  PartitionCount: 1	ReplicationFactor: 1 Configs: 	MarkedForDeletion: true
	Topic: old-version-topic	Partition: 0	Leader: 0	Replicas: 0	Isr: 0	MarkedForDeletion: true

So the topicId of the deleted topic should also remain when deleted, when do you think we should delete the topicId from MetadataCache? suggestions are welcomed.

rajinisivaram replied:
When a topic is deleted, brokers process UpdateMetadataRequest and remove deleted topics from their cache. We track deletion state in ZooKeeper and as you mentioned, you can get this information by directly going to ZK in kafka-topics.sh. But we don't retain that information in every broker. I would remove topic id in the code segment just below this when the partition state is removed from the MetadataCache since we cannot clearly have a map that keeps growing in brokers. Is there a reason why we would want to retain topic id in every broker even after the topic has been deleted? We can't get this information through existing metadata request from brokers anyway. I guess in future, we can add additional metadata to track deleted topic ids if we wanted to, but for now it seems better to delete topic ids from MetadataCache when we delete the partition from the cache.  What do you think?

jolshan replied:
So it seems like there is some logic to remove the partition states of deleted topics from the MetadataSnapshot. Would we want to do something similar there but with topic Ids? Apologies if I'm missing something

rajinisivaram replied:
Yes, that was my suggestion too, but my wording wasn't right. I will reword that to avoid confusion.

dengziming replied:
Thank you for your suggestions, I now understand the solution. I remove the topicId in `removePartitionInfo`.

rajinisivaram reviewed (commented):
@dengziming Thanks for the PR, looks good. Left some comments.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -320,6 +333,7 @@ public String toString() {
             return "TopicMetadata{" +
                 "error=" + error +
                 ", topic='" + topic + '\'' +
+                ", topicId='" + topicId.toString() + '\'' +
```
nit: toString() unnecessary

rajinisivaram commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -223,6 +224,31 @@ class MetadataRequestTest extends BaseRequestTest {
+    assertEquals(2, resp2.topicMetadata.size)
+    resp2.topicMetadata.forEach { topicMetadata =>
+      assertEquals(Errors.NONE, topicMetadata.error)
+      assertNotEquals(Uuid.ZERO_UUID, topicMetadata.topicId())
```
we probably also want to assert that the topic id is not null here (even though we currently never return null).

rajinisivaram reviewed (commented):
@dengziming Thanks for the updates, looks good, just a couple more comments left.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/Cluster.java:
```diff
@@ -327,6 +346,14 @@ public Node controller() {
+        return topicIds.values();
+    }
+
+    public Uuid getTopicId(String topic) {
```
nit: We don't use `get` prefix in other methods. Since this is part of the public API, can we change this to `topicId(String topic)` before merging?

dengziming replied:
Thank you for your reminder, it's helpful to me.

jolshan replied:
Can we also add this and any other new public APIs to the KIP page?

dengziming replied:
Thank you for your reminder, I have added these 2 methods to the KIP page, take a look if you have time.

jolshan replied:
Thanks. looks good!

rajinisivaram reviewed (commented):
@dengziming Thanks for the update, just one minor comment left, apart from that LGTM.

rajinisivaram reviewed (approved):
@dengziming Thanks for the update, LGTM. Will review PR #9626 as well and merge these two together, hopefully later today.

rajinisivaram commented:
There is a compilation failure in the Java 8 build because the PR uses an API not available in Scala 2.12. Will push a fix.

rajinisivaram committed:
Fix Scala 2.12 build

rajinisivaram commented:
@dengziming Thanks for the PR, merging to trunk.
